File: in_place.go
package main

import (
	"fmt"
	"os"
)

func quicksort(a []int) {
	if len(a) <= 1 {
		return
	}
	qsort(a, 0, len(a)-1)
}

func qsort(a []int, lo, hi int) {
	if lo >= hi {
		return
	}
	p := partition(a, lo, hi)
	qsort(a, lo, p-1)
	qsort(a, p+1, hi)
}

func partition(a []int, lo, hi int) int {
	pivot := a[hi]
	i := lo
	for j := lo; j < hi; j++ {
		if a[j] < pivot {
			a[i], a[j] = a[j], a[i]
			i++
		}
	}
	a[i], a[hi] = a[hi], a[i]
	return i
}

func main() {
	verbose := false
	prePrint := false
	for _, arg := range os.Args[1:] {
		if arg == "-v" {
			verbose = true
		} else if arg == "-p" {
			prePrint = true
		}
	}
	xs := make([]int, 10000)
	for i := 0; i < 10000; i++ {
		xs[i] = (i*73 + 19) % 10000
	}
	if prePrint {
		fmt.Println("Before sorting:")
		fmt.Println(xs)
		return
	}
	quicksort(xs)
	if verbose {
		fmt.Println(xs)
	} else {
		fmt.Println(xs[:10])
	}
}

File: in_place.rs
use std::env;

fn quicksort(arr: &mut [i32], lo: isize, hi: isize) {
    if lo >= hi || lo < 0 {
        return;
    }
    
    let p = partition(arr, lo, hi);
    quicksort(arr, lo, p - 1);
    quicksort(arr, p + 1, hi);
}

fn partition(arr: &mut [i32], lo: isize, hi: isize) -> isize {
    let pivot = arr[hi as usize];
    let mut i = lo;
    
    for j in lo..hi {
        if arr[j as usize] < pivot {
            arr.swap(i as usize, j as usize);
            i += 1;
        }
    }
    
    arr.swap(i as usize, hi as usize);
    i
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let verbose = args.contains(&String::from("-v"));
    let pre_print = args.contains(&String::from("-p"));
    
    // Generate test data
    let mut arr = Vec::with_capacity(10000);
    for i in 0..10000 {
        arr.push((i * 73 + 19) % 10000);
    }
    
    if pre_print {
        println!("Before sorting:");
        println!("{:?}", arr);
        return;
    }
    
    let len = arr.len() as isize;
    quicksort(&mut arr, 0, len - 1);
    
    if verbose {
        println!("{:?}", arr);
    } else {
        println!("{:?}", &arr[..10]);
    }
}
